### PR TITLE
Support grouping rules like @media and @layer and fix shadow DOM support

### LINKED ISSUE #41: pseudo-state rules passed to a scss media query mixin do not appear to be added to the DOM
`:hover` definitions defined in scss and passed through to a media query mixin via`@include` and `@content` do not appear to be added to the DOM.

I am using:

├── @storybook/html@6.5.12
├── storybook-addon-pseudo-states@1.15.1

The storybook bits work great, thank you. The `pseudo-class` class is added where I want it, but the related CSS definition is not added to the DOM.

Works:
```
.component:hover {
    background-color: purple;
}
```

Does not work:
```
@include medium {
    .component:hover {
        background-color: purple;
    }
}
```

Where mixins have been previously defined:

```
@mixin medium {
  @include breakpoint(777px) {
    @content;
  }
}

@mixin breakpoint($breakpoint) {
  @media (min-width: #{$breakpoint}) {
    @content;
  }
}
```

I'm not sure if this has to do with the rules being inside a mixin, a media query, or what.

### PR DESCRIPTION
**Pull request by m-akinc**
Fixes #41 and #93

- Added support for `CSSGroupingRule`s like `@layer`, `@media`, etc.
- Fixed bugs in support for shadow DOM.
   - When multiple such elements on the page shared stylesheets, they would not all be added to `shadowHosts` because of the use of `__pseudoStatesRewritten`. This would prevent them from getting classes applied by `shadowHosts.forEach(updateShadowHost)`.
   - When rewriting a selector like `:host(.foo) .control:hover`, it was incorrectly assuming any state selectors (e.g. `:hover`) occurred inside the `:host()` selector and generated a bogus selector like `:host(.foo) .control.pseudo-hover-all`. Now handling scenarios like this by writing `:host(.foo.pseudo-hover-all) .control` instead.
   - `pseudo-*-all` classes were being applied to a custom element (by the `attachShadow` override) even if none of its stylesheets were rewritten (and therefore it was not added to `shadowHosts`). This could result in those classes not being removed later.
   - `updateShadowHost` was only looking at classes on ancestors, so any `pseudo-*` class applied directly to the shadow host because of a selector (i.e. in `applyParameter`) would get removed.
   - `updateShadowHost` would propagate `pseudo-hover`, for example, from an ancestor to a shadow host. It should only be propagating `pseudo-*-all`.
   - `updateShadowHost` wouldn't cross shadow boundaries, so nested shadow DOM (i.e. a custom element in the shadow DOM of a custom element) wouldn't get the global  `pseudo-*-all` class.
- Added support for targeting elements in shadow DOM with selectors:
```js
parameters = {
  pseudo: {
    hover: ":host(.hover) .innerControl"
  }
}
```
- Added new `rewriteStyleSheet` test cases
- Had to update version of tsup due to error I was running into

### CODE REVIEW

**Comment by m-akinc** [comment]
@ghengeveld @JonathanKolnik 🙏 Could I get a review of this PR? I think it would open this addon up to a whole lot of new potential users that can't currently use it (like my team 😁).

**Comment by ghengeveld** [comment]
@m-akinc I'll take a good look at this and your other PRs tomorrow. You did a lot of great work here!

File: src/preview/rewriteStyleSheet.test.ts [hunk 1]
```diff
@@ -107,25 +253,120 @@ describe("rewriteStyleSheet", () => {
...
+  it('rewrites rules inside "@media"', () => {
+    const sheet = new Sheet(
+      `@media (max-width: 790px) {
+        test:hover {
+          background-color: green;
+        }
+      }`
+    )
+    rewriteStyleSheet(sheet as any)
+    const selectors = (sheet.cssRules[0] as GroupingRule).cssRules[0].getSelectors()
```
Comment (ghengeveld) [inline src/preview/rewriteStyleSheet.test.ts:287]: Maybe also assert that `sheet.cssRules[0]` is still the `@media` block?

File: src/preview/withPseudoState.ts [hunk 2]
```diff
@@ -148,7 +174,11 @@ export const withPseudoState: DecoratorFunction = (
 const rewriteStyleSheets = (shadowRoot?: ShadowRoot) => {
   let styleSheets = Array.from(shadowRoot ? shadowRoot.styleSheets : document.styleSheets)
   if (shadowRoot?.adoptedStyleSheets?.length) styleSheets = shadowRoot.adoptedStyleSheets
-  styleSheets.forEach((sheet) => rewriteStyleSheet(sheet, shadowRoot, shadowHosts))
+  let rewroteStyles = false
+  styleSheets.forEach((sheet) => {
+    if (rewriteStyleSheet(sheet, shadowRoot)) rewroteStyles = true
+  })
```
Comment (ghengeveld) [inline src/preview/withPseudoState.ts:180]:
```suggestion
  const rewroteStyles = styleSheets
    .map((sheet) => rewriteStyleSheet(sheet, shadowRoot))
    .some(Boolean)
```

**Review by ghengeveld — APPROVED** [review]
Wow, this is really great work. 👏 I only have a few minor suggestions which you may want to look at.

I would've appreciated some new Storybook stories to showcase the scenarios this PR is addressing, so that we also have visual confirmation that it works as intended (and Chromatic will snapshot them to detect future regressions). Any chance you can add some?

**Comment by m-akinc** [comment]
I added a story for a nested custom component and one for CSS at-rules. Let me know if you need anything else before merging.

**Review by ghengeveld — APPROVED** [review]
LGTM. Nice work!

**Comment by m-akinc** [comment]
> LGTM. Nice work! Ready to merge?

Yes! 🚀

**Comment by ghengeveld** [comment]
CI is failing to release, [working on that](https://github.com/chromaui/storybook-addon-pseudo-states/pull/106)...

**Pull request merged**